src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,352 @@
+            Option<CodeStyleOption<TOptionKind>> option,
+            string descriptorId,
+            LocalizableString title,
+            LocalizableString message)
+        {
+            _option = option;
+            _descriptorId = descriptorId;
+            _title = title;
+            _message = message;
+        }
+
+        #region analysis
+
+        protected abstract DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory();
+        protected abstract bool DiagnosticsForOpenFileOnly(Workspace workspace);
+
+        protected virtual Action<CodeBlockAnalysisContext, TOptionKind> GetCodeBlockAction() => null;
+        protected virtual (ImmutableArray<TSyntaxKind>, Action<SyntaxNodeAnalysisContext, TOptionKind>) GetSyntaxNodeAction() => default;
+
+        private void DiagnosticAnalyzerInitialize(AnalysisContext context)
+        {
+            var analyzeCodeBlock = GetCodeBlockAction();
+            if (analyzeCodeBlock != null)
+            {
+                context.RegisterCodeBlockAction(c => AnalyzeIfEnabled(c, analyzeCodeBlock));
+            }
+
+            var (syntaxKinds, analyzeSyntax) = GetSyntaxNodeAction();
+            if (analyzeSyntax != null)
+            {
+                context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds);
+            }
+
+            // check for all the other things tht can be registered as well.
+        }
+
+        private void AnalyzeIfEnabled(CodeBlockAnalysisContext context, Action<CodeBlockAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext, TOptionKind> analyze, 
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
+
+            analyze(context, optionValue.Value);
+        }
+
+        public abstract class CodeStyle2DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```

Comment:
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that at the bottom of this file.